### PR TITLE
Add "DynamicFunctionReturnTypeExtension" for "array_replace" + fix for "array_merge"

### PR DESCRIPTION
**Pull request by voku**
This Extension is a copy of "ArrayMergeFunctionDynamicReturnTypeExtension"

⇾ requested here: https://github.com/phpstan/phpstan-src/pull/622

INFO: Currently we are getting wrong types from "ArrayMergeFunctionDynamicReturnTypeExtension" because PHP handles array-keys different here: https://3v4l.org/YfKYq
- if the key is numeric, then the value will be added
- if the key is non-numeric, then the value will be replaced

### CODE REVIEW

File: src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		$keyTypes = [];
+		$valueTypes = [];
+		$nonEmpty = false;
+		foreach ($functionCall->getArgs() as $arg) {
+			$argType = $scope->getType($arg->value);
+			if ($arg->unpack) {
+				$argType = $argType->getIterableValueType();
+				if ($argType instanceof UnionType) {
+					foreach ($argType->getTypes() as $innerType) {
+						$argType = $innerType;
```
Comment (ondrejmirtes) [inline src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php:40]: I don't understand this part - the last inner type in the union will win in the `$argType` variable?

Comment (ondrejmirtes) [reply src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php:40]: I know the same code is in ArrayMergeFunctionDynamicReturnTypeExtension - it also puzzles me there.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
How is this different to ArrayMergeFunctionDynamicReturnTypeExtension? Are they going to differ in the future?

File: src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php (hunk 1 above)
Comment (ondrejmirtes) [reply src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php:40]: :)

Comment (ondrejmirtes) [reply src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php:40]: Let's try and see what happens :)

Comment (voku) [reply src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php]: I re-added the fixes and added a limit for the array. I also added the fix for "array_merge" here, so we can see the differences between the implementations in one diff.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Please add a regression test for https://github.com/phpstan/phpstan/issues/5846 here and I'll take a look :) Thank you.

**Comment by voku** [comment]
> Please add a regression test for [phpstan/phpstan#5846](https://github.com/phpstan/phpstan/issues/5846) here and I'll take a look :) Thank you.

I re-added some changes, even more tests and added some checks to prevent performance issues, but maybe we need to adjust the consts "MAX_*" settings? I already tested the performance with a private project and blackfire.io and for me, it looks good, but as airways it depends on the tested code.

**Pull request closed without merging**